### DISCUSSION TITLE
Can't find way to test session expiration in a WebTestCase

### DISCUSSION THREAD

**Discussion by jdevinemt** [symfony/symfony#58005] [Q&A]
I'm trying to test my application's session expiration logic with a WebTestCase, but cannot find a way to do so when session expiration is based on the logic described in [the Symfony session documentation](https://symfony.com/doc/current/session.html#session-idle-time-keep-alive). 

My application code is a little different, but here is a similar example:

```php
<?php

namespace App\EventSubscriber;

use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpKernel\Event\RequestEvent;
use Symfony\Component\HttpKernel\KernelEvents;

class ExampleRequestSubscriber implements EventSubscriberInterface
{
    public static function getSubscribedEvents(): array
    {
        return [
            KernelEvents::REQUEST => [
                ['handleSessionExpiration', 0]
            ]
        ];
    }

    public function handleSessionExpiration(RequestEvent $event): void
    {
        $request = $event->getRequest();

        if(!$request->hasSession()){
            return;
        }

        $session = $request->getSession();

        // lifetime number (# seconds in a day) here would be represented by a SESSION_LIFETIME_SECONDS class constant
        //   in production code
        $expiresAtTimestamp = $session->getMetadataBag()->getLastUsed() + 86400;

        if($expiresAtTimestamp <= time()){
            // invalidate the session and other application-specific logic
        }
    }
}
```

The issue is that there is no good way to create a session with an explicit last used timestamp in a WebTestCase test. The closest I've gotten is via something similar to the following:

```php
<?php

namespace App\Tests\Controller\EventSubscriber;

use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
use Symfony\Component\BrowserKit\Cookie;
use Symfony\Component\HttpFoundation\Session\Storage\MetadataBag;

class RequestEventSubscriberTest extends WebTestCase
{
    /** @test */
    public function session_is_ended_when_expiration_date_is_reached(): void
    {
        $client = static::createClient();

        $session = $client->getContainer()->get('session.factory')->createSession();
        $session->start();
        // trying to set a last updated timestamp that would trigger session expiration logic
        // again, 86400 would be represented by a class constant in production code
        $createdAtTimestamp = time() + 86400 + 1;
        $metadata = [
            MetadataBag::CREATED => $createdAtTimestamp,
            MetadataBag::UPDATED => $createdAtTimestamp
        ];
        $session->getMetadataBag()->initialize($metadata);
        $session->save();

        $client->getCookieJar()->set(new Cookie($session->getName(), $session->getId()));

        $client->request('GET', '/test/blank');

        $sessionAfterRequest = $client->getRequest()->getSession();

        dd($sessionAfterRequest->getMetadataBag()->getLastUsed()); // will output current timestamp
    }
}
```

The issue is that the `initialize()` method of the `MetadataBag` class sets the last updated timestamp to the current time if the last updated timestamp provided minus the current time is less than the update threshold. The update threshold default is 0, and that is how I want it configured for my application. So, I don't necessarily want to have to change the [session.metadata_update_threshold](https://symfony.com/doc/current/reference/configuration/framework.html#metadata-update-threshold) framework config value in my test environment because it wouldn't be testing with a production-like configuration. None of the session services in the container seem to have a method to create a session with a given `MetadataBag`. If they did, I could just manually set the adjustment threshold via its constructor for this specific test.

Currently, I'm just working around this by storing a custom session value to manage the session expiration and updating it on each request. I'd like to use methods Symfony provides though, because it would one less thing I have to manage and is already detailed in their documentation.

Thanks for taking the time to read this :D

**Comment by MatTheCat** [comment]
Never tested this kind of thing, but I see the `NativeSessionStorage` exposes a `setMetadataBag` method; what if you pass it a stub whose `getLastUsed` method returns the value you want?

**Reply by jdevinemt** [reply]
Thank you for the suggestion. I'll look into that solution, although I don't like to mock/stub third party code.

**Reply by jdevinemt** [reply]
Actually, your answer gave me an idea for a solution, which is probably better than what I was trying to accomplish anyway. Instead of trying to manipulate the time the session was last used at, which is controlled by third party code, I should instead manipulate the date that is being compared against (in a real use case, the current timestamp). I already have a `DateTimeProviderInterface` for this purpose when I need to make assertions against "now". On production it uses a real time `DateTimeProvider` class whose `now()` method returns the current datetime. In the test environment it uses a `FixedDateTimeProvider` class whose `now()` method returns a fixed datetime that is initialized on construction and only changes when requested via a `reset()` method. I may post my full solution if it ends up working and I have time today.

**Reply by MatTheCat** [reply]
FYI it seems you implemented something similar to the clock component: https://symfony.com/doc/current/components/clock.html

**Reply by jdevinemt** [reply]
Brilliant! I did not know that existed.

**Reply by jdevinemt** [reply]
Dang.. I am still on Symfony 5.4 and it looks like it isn't available until 6.2. This test is being written as part of a migration to a new platform and eventually a newer version of Symfony. I'll just have to use my custom solution until we get onto a newer Symfony version.
